mlflow/utils/promptlab_utils.py

```diff
@@ -0,0 +1,143 @@
+    """
+    return conda_yaml.strip()
+
+
+def create_python_env_file() -> str:
+    python_yaml = """
+    dependencies:
+      - -r requirements.txt
+    """
+    return python_yaml.strip()
+
+
+def create_requirements_txt_file(mlflow_version: str) -> str:
+    requirements_content = f"""
+    mlflow[gateway]=={mlflow_version}
+    """
+    return requirements_content.strip()
+
+
+def create_model_file(
+    run_uuid: str,
+    mlflow_version: str,
+    prompt_parameters: List[Param],
+    model_uuid: str,
+    utc_time_created: str,
+) -> str:
+    inputs_signature = [{"name": param.key, "type": "string"} for param in prompt_parameters]
+
+    outputs_signature = [{"name": "output", "type": "string"}]
+
+    model_file = {
+        "artifact_path": "model",
+        "flavors": {
+            "python_function": {
+                "env": {"conda": "conda.yaml", "virtualenv": "python_env.yaml"},
+                "loader_module": "gateway_loader_module",
+                "code": "loader",
+            }
+        },
+        "mlflow_version": mlflow_version,
+        "model_uuid": model_uuid,
+        "run_id": run_uuid,
+        "utc_time_created": utc_time_created,
+        "metadata": {"mlflow_uses_gateway": "true"},
+        "saved_input_example_info": {
+            "artifact_path": "input_example.json",
+            "type": "dataframe",
+            "pandas_orient": "split",
+        },
+        "signature": {
+            "inputs": json.dumps(inputs_signature),
+            "outputs": json.dumps(outputs_signature),
+        },
+    }
+
+    model_json = json.dumps(model_file)
+    return model_json
+
+
+def create_input_example_file(prompt_parameters: List[Param]) -> str:
+    input_example = {"inputs": [param.value for param in prompt_parameters]}
+    input_example_json = json.dumps(input_example)
+    return input_example_json
+
+
+def create_loader_file(prompt_parameters, prompt_template, model_parameters, model_route):
+    python_inputs = ",\n\t\t\t\t".join(
+        [f"{param.key}=inputs['{param.key}'][idx]" for param in prompt_parameters]
+    )
+
+    # Replace {{parameter}} with $parameter in the prompt template
+    python_template = prompt_template.replace("{{", "$").replace("}}", "")
+
+    # Escape triple quotes in the prompt template
+    sanitized_python_template = python_template.replace('"""', '\\"""')
+
+    python_parameters = ",\n\t\t\t\t\t".join(
+        [f'"{param.key}": {param.value}' for param in model_parameters]
+    )
+
+    loader_module_text = f"""
+# loader/gateway_loader_module.py
+from typing import List, Dict
+from string import Template
+import pandas as pd
+import mlflow.gateway
+
+mlflow.gateway.set_gateway_uri("databricks")
+
+class GatewayModel:
+    def __init__(self, model_path):
+        self.prompt_template = Template(\"\""
+{sanitized_python_template}\n\""")
+
+    def predict(self, inputs: pd.DataFrame) -> List[str]:
+        results = []
+        for idx in inputs.index:
```

Comment:
Can we create a template as a python file, read it, and replace `__xx___`?

```python
# loader/gateway_loader_module.py
from typing import List, Dict
from string import Template
import pandas as pd
import mlflow.gateway

mlflow.gateway.set_gateway_uri("databricks")


class GatewayModel:
    def __init__(self, model_path):
        self.prompt_template = Template(
            """
__sanitized_python_template__
"""
        )

    def predict(self, inputs: pd.DataFrame) -> List[str]:
        results = []
        for idx in inputs.index:
            prompt = self.prompt_template.substitute(__python_inputs__)
            result = mlflow.gateway.query(
                route="{model_route}",
                data={
                    {
                        "prompt": prompt,
                        **__python_parameters__,
                    }
                },
            )
            results.append(result["candidates"][0]["text"])
        return results


def _load_pyfunc(model_path):
    return GatewayModel(model_path)

```

Reply:
Yea I think this is a great idea. I push commits that use a template txt file. I am not sure how to make a python file with these template variables.